cannon/mipsevm/mips.go

```diff
@@ -285,13 +285,13 @@ func (m *InstrumentedState) mipsStep() error {
 
 	// j-type j/jal
 	if opcode == 2 || opcode == 3 {
-		// TODO likely bug in original code: MIPS spec says this should be in the "current" region;
-		// a 256 MB aligned region (i.e. use top 4 bits of branch delay slot (pc+4))
 		linkReg := uint32(0)
 		if opcode == 3 {
 			linkReg = 31
 		}
-		return m.handleJump(linkReg, SE(insn&0x03FFFFFF, 26)<<2)
+		// Take top 4 bits of current PC (its 256 MB region), and concatenate with 26-bit offset
+		target := (m.state.NextPC & 0xF0000000) | ((insn & 0x03FFFFFF) << 2)
+		return m.handleJump(linkReg, target)
```

Comment:
Most emulators actually get this wrong, or completely ignore the top 4 bits, but I believe we got it correct here now. It shouldn't be the 4 bits of the current PC, but the 4 bits of the next PC (delay slot), or else the "programming notes" in the spec wouldn't hold true. The spec expression of this operation is quite unclear about it though.

> This definition creates the following boundary case: When the branch instruction is in the last word of a 256MB
region, it can branch only to the following 256MB region containing the branch delay slot.

So because it's using next-PC, the top 4 bits will match that of the next 256MB (`2**28`, the remaining bits) region.

And it's also correctly not doing sign-extension (`SE`) anymore, as the spec says "PC-region branch (not PC-relative);".

Reply:
yup, meant to say nextPC in the comments based on this bit in the spec:
> This definition creates the following boundary case: When the branch instruction is in the last word of a 256 MB
region, it can branch only to the following 256 MB region containing the branch delay slot.

Also, the program counter in architecture documents, unlike the way we use it in the VM, typically refers to next instruction, i.e. `nextPC`.
Will fix the comments.